extensions/amp-analytics/0.1/crypto-impl.js

```diff
@@ -14,51 +14,72 @@
  * limitations under the License.
  */
 
-import * as lib from '../../../third_party/closure-library/sha384-generated';
 import {fromClass} from '../../../src/service';
 import {dev} from '../../../src/log';
+import {getExistingServiceForWindow} from '../../../src/service';
+import {extensionsFor} from '../../../src/extensions';
 import {stringToBytes} from '../../../src/utils/bytes';
+import {base64UrlEncodeFromBytes} from '../../../src/utils/base64';
 
 /** @const {string} */
 const TAG = 'Crypto';
 const FALLBACK_MSG = 'SubtleCrypto failed, fallback to closure lib.';
 
 export class Crypto {
 
+  /**
+   * @param {!Window} win
+   */
   constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
     /** @private @const {?webCrypto.SubtleCrypto} */
     this.subtle_ = getSubtle(win);
+
+    /** @private {?Promise<{sha384: function((string|Uint8Array))}>} */
+    this.polyfillPromise_ = null;
+
+    if (!this.subtle_) {
+      this.loadPolyfill_();
+    }
   }
 
   /**
    * Returns the SHA-384 hash of the input string in a number array.
    * Input string cannot contain chars out of range [0,255].
    * @param {string|!Uint8Array} input
-   * @returns {!Promise<!Array<number>>}
+   * @returns {!Promise<!Uint8Array>}
    * @throws {!Error} when input string contains chars out of range [0,255]
    */
   sha384(input) {
-    if (this.subtle_) {
-      try {
-        return this.subtle_.digest({name: 'SHA-384'},
-                input instanceof Uint8Array ? input : stringToBytes(input))
-            // [].slice.call(Unit8Array) is a shim for Array.from(Unit8Array)
-            /** @param {?} buffer */
-            .then(buffer => [].slice.call(new Uint8Array(buffer)),
-                e => {
-                  // Chrome doesn't allow the usage of Crypto API under
-                  // non-secure origin: https://www.chromium.org/Home/chromium-security/prefer-secure-origins-for-powerful-new-features
-                  if (e.message && e.message.indexOf('secure origin') < 0) {
-                    // Log unexpected fallback.
-                    dev().error(TAG, FALLBACK_MSG, e);
-                  }
-                  return lib.sha384(input);
-                });
-      } catch (e) {
-        dev().error(TAG, FALLBACK_MSG, e);
-      }
+    if (!(input instanceof Uint8Array)) {
```

Comment:
Each `window` has it's own instance of the generic classes, take `String` for instance. Now, when we create an iframe, it gets its own `window` and its own `String`. If we compare, they are not equivalent:

```js
window.String === iframe.contentWindow.String; // => false
```

Since `instanceof` looks for object equality in the prototype chain, this also breaks across iframe boundaries:

```js
iframe.contentWindow.uint8instance instanceof Uint8Array; // => false
uint8instance instanceof iframe.contentWindow.Uint8Array; // => false